process/marshal.go

```diff
@@ -232,6 +236,24 @@ func (prevote Prevote) MarshalBinary() ([]byte, error) {
 	if err := binary.Write(buf, binary.LittleEndian, prevote.blockHash); err != nil {
 		return buf.Bytes(), fmt.Errorf("cannot write prevote.blockHash: %v", err)
 	}
+	if err := binary.Write(buf, binary.LittleEndian, uint64(len(prevote.nilReasons))); err != nil {
+		return buf.Bytes(), fmt.Errorf("cannot write prevote.nilReasons len: %v", err)
+	}
+	for key, val := range prevote.nilReasons {
+		keyBytes := []byte(key)
+		if err := binary.Write(buf, binary.LittleEndian, uint64(len(keyBytes))); err != nil {
+			return buf.Bytes(), fmt.Errorf("cannot write prevote.nilReasons key len: %v", err)
+		}
+		if err := binary.Write(buf, binary.LittleEndian, keyBytes); err != nil {
+			return buf.Bytes(), fmt.Errorf("cannot write prevote.nilReasons key data: %v", err)
+		}
+		if err := binary.Write(buf, binary.LittleEndian, uint64(len(val))); err != nil {
+			return buf.Bytes(), fmt.Errorf("cannot write prevote.nilReasons val len: %v", err)
+		}
+		if err := binary.Write(buf, binary.LittleEndian, val); err != nil {
+			return buf.Bytes(), fmt.Errorf("cannot write prevote.nilReasons val data: %v", err)
+		}
+	}
```

Comment:
To ensure that two prevotes with the same data will result in the same hash, we need to iterate over the nilReasons in a deterministic way. Go maps are iterated over non-deterministically, so we should convert it to a sorted list first.

Reply:
We should have a test case that captures this error (i.e. marshal random messages multiple times and check that the bytes are the same, check that random messages always have the same hash, etc.). This will help protect these bugs from potentially slipping through in the future.